QgsAttributeTableModel: Moved feature loader into its own thread

elpaso opened the pull request:
Attempt to fix #12926

This is an attempt to keep the GUI responsive while loading large datasets into the attribute table, it replaces the current QProgressDialog implementation by moving the loader in its own thread and adding a progress bar and abort button in the main dialog.

The original idea is from @jef-n that suggested me a threaded approach in Las Palmas.

My thread programming experience is close to 0 so the implementation might be buggy but according to my quick tests on 1^E6 tables, it behaves pretty well.

Comments welcome!

![qgis_attrtable_worker](https://cloud.githubusercontent.com/assets/142164/11476781/e53c99aa-9784-11e5-837b-64264ae39585.png)

elpaso committed:
Moved feature loader into its own thread

Attempt to fix #12926

m-kuhn commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -87,6 +87,8 @@ QgsAttributeTableDialog::QgsAttributeTableDialog( QgsVectorLayer *theLayer, QWid
 {
   setupUi( this );
 
+  mLoadAborted =  FALSE;
```
You should use `false` instead

elpaso replied:
Ok, thanks!

m-kuhn commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -857,3 +865,32 @@ void QgsAttributeTableDialog::setFilterExpression( const QString& filterString )
+
+void QgsAttributeTableDialog::on_mLoadAbortButton_clicked()
+{
+  mLoadAborted = TRUE;
```
Dito

m-kuhn replied:
up

m-kuhn commented on src/app/qgsattributetabledialog.cpp:
```diff
@@ -857,3 +865,32 @@ void QgsAttributeTableDialog::setFilterExpression( const QString& filterString )
+{
+  mLoadProgressBar->setFormat( "%v / %m" );
+  mLoadProgressBar->setMaximum( numFeatures );
+  mLoadAborted = FALSE;
```
Dito

m-kuhn commented on src/gui/attributetable/qgsattributetableloadworker.cpp:
```diff
@@ -0,0 +1,69 @@
+
+QgsAttributeTableLoadWorker::QgsAttributeTableLoadWorker( const QgsFeatureIterator &features ):
+    mIsRunning( FALSE ),
+    mStopped( FALSE )
```
Dito

m-kuhn commented on src/gui/attributetable/qgsattributetableloadworker.cpp:
```diff
@@ -0,0 +1,69 @@
+
+void QgsAttributeTableLoadWorker::stopJob()
+{
+  mStopped = TRUE;
```
...

m-kuhn commented on src/gui/attributetable/qgsattributetableloadworker.cpp:
```diff
@@ -0,0 +1,69 @@
+{
+  QgsDebugMsg( "QgsAttributeTableLoadWorker started!" );
+  mStopped = FALSE;
+  mIsRunning = TRUE;
```
Ok, I'll stop here, you find them yourself :)

m-kuhn commented on src/gui/attributetable/qgsattributetablemodel.cpp:
```diff
@@ -40,6 +41,7 @@ QgsAttributeTableModel::QgsAttributeTableModel( QgsVectorLayerCache *layerCache,
     , mLayerCache( layerCache )
     , mFieldCount( 0 )
     , mCachedField( -1 )
+    , mLoadWorker( NULL )
```
If a null representation should be used in C++ it's better to use `nullptr` which is actually defined in the standard.

jef-n replied:
No, that doesn't work on old compilers.   0 is fine.

m-kuhn replied:
That would be easy to`#define` just like `override`. But it has the advantage that on new compilers it can't be compared to integers by accident. While NULL doesn't offer any advantages over plain 0 nowhere.

m-kuhn commented on src/gui/attributetable/qgsattributetablemodel.cpp:
```diff
@@ -340,10 +351,70 @@ void QgsAttributeTableModel::loadAttributes()
+  QgsFeatureIterator features = mLayerCache->getFeatures( mFeatureRequest );
+
+  // Pass the total number of features as a maximum for the progress bar
+  long fc = mLayerCache->layer()->featureCount();
```
Is the new API for the iterator count not yet ready? Lost the PR somehow...

elpaso replied:
Not sure, I also searched for a different "count" method but couldn't find anything better.

m-kuhn commented on src/gui/attributetable/qgsattributetablemodel.cpp:
```diff
@@ -340,10 +351,70 @@ void QgsAttributeTableModel::loadAttributes()
+  emit loadStarted( fc );
+
+  // Ensure signals are delivered
+  qApp->processEvents();
```
I was hoping we finally get rid of that with threads. Are you sure that's required?

elpaso replied:
I might be wrong but I believe that the approach I'm using (one of the simplest available) uses signals for iter-thread communication/synchronization and processEvents is is called in the GUI thread and process all signals.
Here it is called to make sure that loadStarted is processed.

m-kuhn replied:
If all the expensive work is done in a separate thread the gui thread will soon be back to the main event loop and therefore there is no need to explicitly trigger this (dangerous) command.

Apart from this as long as a connection is not connected between different threads or not explicitly requested to be done to be queued it will be called directly and not via the eventloop IIRC.

elpaso replied:
Testing now: it works fine without that line. I'll remove it.

elpaso replied:
Just curious: why it's dangerous?

m-kuhn replied:
I think there are other occurrences in the attribute table which were previously necessary and no longer are with this patch.

It is dangerous because it executes code in a context which it was normally not designed to be run in.
E.g. a user action could lead to deletion of an object in the call stack.

m-kuhn commented:
Good stuff, looking forward to this.
Has been on my todo list for too long, glad to see you tackled it :)

Next will be to allow replacing the local filtering with the filter model with provider side filtering. In combination with this and the expression compiler that will be really nice to interactively explore large datasets.

m-kuhn commented:
@wonder-sk any feedback on the multithreading stuff?

m-kuhn commented:
@elpaso please have a look at the tests?

michaelkirk commented on src/app/qgsattributetabledialog.h:
```diff
@@ -69,44 +69,77 @@ class APP_EXPORT QgsAttributeTableDialog : public QDialog, private Ui::QgsAttrib
+
+    /**
+     * @brief Called while features are being loaded, cancel to abort
+     * @param i
```
Can you include a comment as to what *i* is? I think it's the "percentDone" right or is it the count?

Or you could save yourself a comment by calling the var "percentDone" instead of "i".

elpaso replied:
Yes, sure (that code is not mine: it was already there from the old progress dialog).

michaelkirk commented on src/gui/attributetable/qgsattributetablemodel.h:
```diff
@@ -241,9 +249,22 @@ class GUI_EXPORT QgsAttributeTableModel: public QAbstractTableModel
+     * Called while the features are loaded
+     * @note not available in python bindings
+     */
+    void loadProgress( int i, bool &cancel );
```
ditto for this `i`.

michaelkirk commented:
Really happy to see this! Thanks for your work.

elpaso commented:
@m-kuhn 
qgis_dualviewtest fails because now that's threaded, the features are not yet available (in the main thread) when previously they where, not sure about what to do... insert a sleep?

m-kuhn commented:
@elpaso I'd prefer not to sleep, the model should emit something when it has finished loading, so better wait for that.

elpaso committed:
Modified tests to wait for thread loader

elpaso commented:
@m-kuhn , @michaelkirk thanks for your comments so far, I've corrected the issues (I hope I didn't forget anything) and rewrote some tests. Any help with testing this patch and to make it better would be highly appreciated!

nyalldawson commented on src/gui/attributetable/qgsattributetableloadworker.cpp:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+  // Remaining features?
+  if ( i % 1000 )
```
Would be safer to check 

    if ( !features.isEmpty() )

elpaso replied:
Why? 
If `features` is empty, the `Q_FOREACH` in `QgsAttributeTableModel::featuresReady` slot will just do nothing.
Or maybe the check should be in that slot?

nyalldawson replied:
I just meant it avoids having the hardcoded 1000 in there twice... it's possible that in future someone will change this constant, and it's better if there's only one place required to do that.

elpaso replied:
Ok. Good advice! Thanks.

michaelkirk commented:
One thought while this compiles, from a UX consistency standpoint, I think progress bars are usually on the bottom. Specifically, in the main QGIS window, it's in the bottom left. This is also consistent with where Chrome puts it's progress/activity indicator bar. Do you have any counter examples?

Top right might be fine in this case, but just wanted to give it some thought. Is there a reason that you chose top right?

nyalldawson commented:
@elpaso great work... This will make a huge difference to usability of the table!

@m-kuhn I realise this is a bit off topic, but I've always wondered why the entire layer has to be fetched in advance to display the table. Wouldn't it be possible to only retrieve features as required by the model? (I haven't looked deep into this code, I'm just wondering if there's a technical reason why this wouldn't be possible).

elpaso commented:
@michaelkirk to be honest, I didn't spend much time on the GUI design and focus on the threading implementation but this was my quick reasoning:
- the progressbar and the button are not always visible and most of the time they will disappear on less than 1 sec
- positions: top or bottom
- avoid window resize
- find an empty space in the existing top/bottom
the latter led me to choose the top, because the bottom could be occupied by the filter lineedit and the progressbar would trigger a resize of the lineedit (not on the window, that would be worse).

That said, it's perfectly fine for me to move it.

m-kuhn commented:
@nyalldawson short follow up for offtopic: It has been loading everything into ram to block the interface only once and not with every filter change. Especially since the filters were all evaluated in QGIS and so the whole dataset had to be loaded on filter and order by change.
With this patch, the expression compiler (and upcoming order by) I am confident that we can change this.

Although I think that it depends a lot on the scenario which method is preferable: how big the dataset is, if your desktop machine has enough ram to cache everything, if the provider can handle filtering / ordering of a particular query, latency, random access to individual features.

elpaso committed:
Added debug messages and thread stopper method

elpaso commented:
There are still some random segfaults with `ctest  -V -R PyQgsAttributeTableModel`. I definitely need some help with debugging this.

elpaso commented:
Hopefully solved the segfault, now fighting with another (old) bug unrelated with this PR but that shows up here.

elpaso committed:
Rebased with latest commits

elpaso committed:
Removed debug stuff (and warning)

elpaso committed:
Added C++ test for QgsAttributeTableModel

elpaso committed:
Added mutex for thread safety

elpaso committed:
Added wait loader to block request while loading

This is useful while testing and should
be never hit the normal usage but won't hurt anyway.

elpaso committed:
Use wait loader instead of wait signals

elpaso committed:
Added missing signals

elpaso committed:
Added missing signals

elpaso committed:
TRUE -> true

elpaso committed:
Loading slots are now private

elpaso committed:
Removed loadProgress from bindings

elpaso commented:
@m-kuhn, @wonder-sk ready for rewiew: travis is happy and the additional stress test I've added doesn't segfault.

elpaso committed:
Added more locks to critical code

Moved some internal methods to private and
added locks to all cache and layer access.
This should have removed last random race conditions.

elpaso committed:
Default to table view

elpaso committed:
Delay filter model until loading has finished

elpaso commented:
@m-kuhn can you have a look to my solution to delay `generateListOfVisibleFeatures` in https://github.com/elpaso/QGIS/commit/2af87c1968e551bda455c7e1945659f2b5d4d1ee ? I'm not fully convinced, there must be a smarter solution. The patch is now working fine, but filtering on features while they are still loading was still freezing the GUI (because it's done in the GUI thread).
BTW, the GUI behavior with limited number of features is nearly the same as in current master, the main advantage with this patch is that you can start exploring the features table while still loading.
I start thinking that the threaded loader, could be used in the filter model too.

m-kuhn commented:
Why do you need to wait for the features to be loaded? Is it not enough to let the features load and apply the filter condition after the list of visible features has been built in parallel? Or is it mandatory that the filter is set after the master model is populated?
Apart from this I would say you need to disconnect the signal again once you have generated the list.

------------

I would not invest into a solution with an async filter but rather remove the filter model completely in a future version (or only use it for sorting) and set the filter directly on the cache.

Whenever the filter changes, the cache is emptied and rebuilt. Any ongoing loading aborted and restarted with the new filter conditions. This will be much smarter and faster with the expression compiler which we have now and scales for big datasets on modern databases.

elpaso commented:
@m-kuhn thanks for your comments, I'll think about retiring the filter model in favour of direct filtering in the cache, that is much a simpler approach.
The problem with filtering or sorting while features are being loaded from the loader thread is that there are some function calls that iterate through the cache from the GUI thread (`QgsAttributeTableModel::prefetchColumnData` and `QgsAttributeTableFilterModel::generateListOfVisibleFeatures`) freezing the application and vanishing the effect of the threaded loader (keep in mind that we are processing millions of records! In case of small layers, this freeze is unnoticeable).
I only see two possibilities here: 
* disable filtering or sorting until the loader thread has finished (not my preferred one)
* abort the loader and restart it whenever the user sort/filters the data

I guess that the purpose of the cache and the proxy model is to speed up those sorting and filtering operations without fetching data from the provider again but since the filter is applied to the cache when it is created, in most situations it has to be cleared and rebuilt when the filter changes.
A particular case is when the filter is a spatial one (like "show features visible on the map"), in this case, there is no need to clear the cache and the filtering is done using the renderer.

elpaso committed:
Process events before emitting loadStart

m-kuhn commented:
> retiring the filter model in favour of direct filtering in the cache
that's a typo, you wanted to say provider, right?

Theoretically `QgsAttributeTableModel::prefetchColumnData` and `QgsAttributeTableFilterModel::generateListOfVisibleFeatures` could also be executed in a separate thread, with some additional synchronisation logic. But I think the above method will be much easier to implement.

Maybe it's worth having both implementations (cached with filter model and provider side filtering) available side-by-side. For file-based data the caching may be better while for databases (with expression compiling) provider side filtering may be better.

haubourg commented:
@m-kuhn hi Matthias, Is that PR a good candidate for speed improvements on attribute table performance discussed yesterday? Is it still compatible with QGIS 3.0 context?

m-kuhn commented:
Hi Régis,

Yes, that's exactly what I was referring to. I think it needs to be
reviewed (I remember I wasn't sure at that time that there are no
deadlocks possible the way it was done) and then filters and sorting
should be moved to server (I think that's missing from the PR).

But: I just realized that I already moved loading of child features to a
lazy loading (so only when you switch tabs and they become visible). The
main issue are probably not the attribute tables but the relation
reference (or value relation) lists.

Cheers
Matthias

On 01/19/2017 04:00 PM, Régis Haubourg wrote:
> @m-kuhn <https://github.com/m-kuhn> hi Matthias, Is that PR a good
> candidate for speed improvements on attribute table performance
> discussed yesterday? Is it still compatible with QGIS 3.0 context?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/2518#issuecomment-273798375>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAj6d4kWy1ffeejfWEiY7a3gsnRECgo8ks5rT3qegaJpZM4GrjSJ>.
>
